### PR TITLE
mavlink: allow up to 6 instances

### PR DESCRIPTION
**Pull request by dagar**

### CODE REVIEW

**Comment by julianoes** [comment]
This is going to conflict with my gimbal PR, isn't it? :sob:

**Comment by dagar** [comment]
> This is going to conflict with my gimbal PR, isn't it?

Only the last small commit https://github.com/PX4/PX4-Autopilot/pull/16830/commits/e9d33046ee22b7444fdab893fa84443b7cff06c1.

**Comment by dagar** [comment]
Rebased on master after the Gimbal PR has merged.

Although this PR no longer extends mavlink to support 6 instances I'd still like to get the rest in.

File: src/modules/mavlink/mavlink_main.cpp [hunk 1]
```diff
@@ -76,84 +77,15 @@
 #define MAX_DATA_RATE                  10000000        ///< max data rate in bytes/s
 #define MAIN_LOOP_DELAY                10000           ///< 100 Hz @ 1000 bytes/s data rate
 
-static Mavlink *_mavlink_instances = nullptr;
+pthread_mutex_t mavlink_module_mutex = PTHREAD_MUTEX_INITIALIZER;
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_main.cpp:80]:
```suggestion
static pthread_mutex_t mavlink_module_mutex = PTHREAD_MUTEX_INITIALIZER;
```

File: src/modules/mavlink/mavlink_main.cpp [hunk 2]
```diff
@@ -76,84 +77,15 @@
...
-	Mavlink *m = Mavlink::get_instance(channel);
-
-	if (m != nullptr) {
-		return m->get_buffer();
-
-	} else {
-		return nullptr;
-	}
-}
+void mavlink_send_uart_bytes(mavlink_channel_t chan, const uint8_t *ch, int length) { mavlink_module_instances[chan]->send_bytes(ch, length); }
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_main.cpp:84]: Considerably nicer

File: src/modules/mavlink/mavlink_main.cpp [hunk 3]
```diff
@@ -2156,6 +2079,13 @@ Mavlink::task_main(int argc, char *argv[])
 
 #endif // MAVLINK_UDP
 
+	if (!set_instance_id()) {
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_main.cpp:2082]: I haven't checked in detail, but I believe this needs to be after `configure_streams_to_default`

**Comment by dagar** [comment]
Are we good to go here?

**Review by LorenzMeier — APPROVED** [review]

**Pull request merged**